bitcoin/src/blockdata/witness.rs

```diff
@@ -194,25 +233,29 @@ impl Witness {
     pub fn clear(&mut self) {
         self.content.clear();
         self.witness_elements = 0;
-        self.last = 0;
-        self.second_to_last = 0;
+        self.indices_start = 0;
     }
 
     /// Push a new element on the witness, requires an allocation
     pub fn push<T: AsRef<[u8]>>(&mut self, new_element: T) {
         let new_element = new_element.as_ref();
         self.witness_elements += 1;
-        self.second_to_last = self.last;
-        self.last = self.content.len();
+        let previous_content_end = self.indices_start;
         let element_len_varint = VarInt(new_element.len() as u64);
         let current_content_len = self.content.len();
+        let new_item_total_len = element_len_varint.len() + new_element.len();
         self.content
-            .resize(current_content_len + element_len_varint.len() + new_element.len(), 0);
-        let end_varint = current_content_len + element_len_varint.len();
+            .resize(current_content_len + new_item_total_len + U32_SIZE, 0);
+
+        self.content[previous_content_end..].rotate_right(new_item_total_len);
+        self.indices_start += new_item_total_len;
+        encode_cursor(&mut self.content, self.indices_start, self.witness_elements - 1, previous_content_end);
+
+        let end_varint = previous_content_end + element_len_varint.len();
         element_len_varint
-            .consensus_encode(&mut &mut self.content[current_content_len..end_varint])
+            .consensus_encode(&mut &mut self.content[previous_content_end..end_varint])
             .expect("writers on vec don't error, space granted through previous resize");
-        self.content[end_varint..].copy_from_slice(new_element);
+        self.content[end_varint..end_varint + new_element.len()].copy_from_slice(new_element);
```

Comment:
The code seem to be more complicated than needed. I'd write:

```rust
let element_len_varint = VarInt(new_element.len() as u64);
 let new_item_total_len = element_len_varint.len() + new_element.len();
self.content.reserve(new_item_total_len + 4);
let mut varint_buf = [u8; 9];
element_len_varint.consensus_encode(&mut &mut varint_buf)
                .expect("varint is at most 9B long");
self.content.splice(self.indices_start..self.indices_start, varint_buf[..element_len_varint.len()].iter().chain(new_element).copied());
self.content.extend_from_slice(&u32::try_from(self.indices_start).unwrap().to_ne_bytes());
self.indices_start += new_item_total_len;
self.witness_elements += 1;
```

Which also should be faster (no zeroing of vec).